x-pack/filebeat/module/virustotal/livehunt/_meta/fields.yml

```diff
@@ -0,0 +1,924 @@
+        List of representative attributes according to VirusTotal
+      type: keyword
+    - name: trid
+      default_field: false
+      description: >
+        TrID is a utility designed to identify file types from their binary signatures. It may give several detections,
+        ordered by higher to lower probability of file format identification.
+      release: beta
+      type: group
+      fields:
+        - name: file_type
+          description: Identified file type
+          type: keyword
+        - name: probability
+          description: >
+            Probability for a positive identification
+          format: percent
+          type: float
+    - name: type_description
+      description: >
+        Describes the file type
+      type: keyword
+    - name: type_tag
+      description: >
+        Tags representing the file type
+      type: keyword
+- name: file.hash.ssdeep
+  default_field: false
+  description: >
+    Fuzzy hash of the file using ssdeep.
+  release: beta
+  type: keyword
+
+- name: file.elf
+  default_field: false
+  description: >
+    ELF events from VirusTotal Intelligence Live Hunt results.
+  overwrite: true
+  type: group
+  release: beta
+  fields:
+    - name: creation_date
+      default_field: false
+      description: >
+        extracted when possible from the file's metadata. Indicates when it was
+        built or compiled. It can also be faked by malware creators.
+      type: date
+    - name: header
+      default_field: false
+      description: >
+        Header information of the ELF file.
+      release: beta
+      type: group
+      fields:
+        - name: class
+          description: >
+            Header class of the ELF file.
+          type: keyword
+        - name: data
+          description: >
+            Data table of the ELF header.
+          type: keyword
+        - name: machine
+          description: >
+            Machine architecture of the ELF header.
+          type: keyword
+        - name: os_abi
+          description: >
+            NEED TO ADD
+          type: keyword
+        - name: type
+          description: >
+            Header type of the ELF file.
+          type: keyword
+        - name: version
+          description: >
+            Version of the ELF header.
+          type: keyword
+        - name: abi_version
+          type: keyword
+          description: >
+            Version of the ELF Application Binary Interface (ABI).
+        - name: entrypoint
+          format: string
+          type: long
+          description: >
+            Header entrypoint of the ELF file.
+        - name: object_version
+          type: keyword
+          description: >
+            "0x1" for original ELF files.
+
+    - name: number_program_headers
+      description: >
+        Number of ELF Program Headers.
+      type: long
+    - name: number_section_headers
+      description: >
+        Number of ELF Section Headers.
+      type: long
+    - name: sections
```

Comment:
I'm not entirely sure given that virus total returns information about whether artifacts are malicious to begin with, but I imagine that entropy calculations and or hashes might be useful to retain here.

Reply:
Yes, the intention is to keep all the info for the time being. If users don't want a particular fieldset, it can be dropped in the filebeat config or ingest processor. The section data has chi2 calculations and entropy. Virustotal _doesn't_ provide an overall status of malicious or benign, but offers community votes of that, and individual engine assessments

Reply:
similar to comment about imported symbols below, we could normalize section data with something like:

Reply:
After working through several examples and reading on the various binary executable formats, I've come up with this. Thoughts @andrewstucki ?

```yaml
// Abstract structure for all binary types, missing fields for a given data source will be excluded
{
    name: "[keyword] Name of code section",
    physical_offset: "[keyword] Offset of the section from the beginning of the segment, in hex",
    physical_size: "[long] Size of the code section in the file in bytes",
    virtual_address: "[keyword] relative virtual memory address when loaded",
    virtual_size: "[long] Size of the section in bytes when loaded into memory",
    flags: "[keyword] List of flag values as strings for this section",
    type: "[keyword] Section type as string, if applicable",
    segment_name: "[keyword] Name of segment for this section, if applicable"
}


// Mach-O example
{
    file.macho.sections: [
        {
            name: "__nl_symbol_ptr",
            flags: ["S_8BYTE_LITERALS"],
            type: "S_CSTRING_LITERALS",
            segment_name: "__DATA"
        }, ...
    ]
}
// ELF example
{
    file.elf.sections: [
        {
            name: ".data",
            physical_offset: "0x3000",
            physical_size: 16,
            virtual_address: "0x4000",
            flags: ["WA"], // This is how VT presents the data. Pretty sure this maps to ["WRITE", "ALLOC"], but I don't have an exhaustive mapping
            type: "PROGBITS"
        }, ...
    ]
}

// PE example

{
    file.pe.sections: [
        {
            name: ".data",
            physical_size: 2542592,
            virtual_address: "0x2DE000",
            virtual_size: 2579264,
            flags: ["rw"], // Again, this is how VT presents it. Likely maps to ["MEM_READ", "MEM_WRITE"], but I don't have an exhaustive mapping
            type: ".data",
            entropy: 6.83,
            chi2: 13360996
        }, ...
    ]
}
```

I'm least pleased by my Mach-O example, but I think that's mostly limited to how VT provides the data currently. It provides offset info for each segment, and then lists sections that exist within the segment with no info at all. This is the only reason, I think to even mention the segment name, though that could be omitted and and be listed within each segment data as a list of included sections.

Finally, I think this at least works for a common fieldset of section data. The flags we can improve over time since it will be a list of keywords, and for PE, I think it's hard-coded as an attribute of the section name/type.

Reply:
@dcode that looks pretty similar to what I was thinking. Thinking through the `flags` bit trips me up a little too--I'm thinking that eventually we may want some verbiage in the description that says something to the effect of "use whatever constant name is found in the spec/OS headers" >_>. If we wanted to be strict about it, a VT filebeat module could always just normalize the VT payload to whatever we wanted.

Also, for reference,  sections do have offset and size info associated with them, so despite the VT api shortcomings, pretty sure the same fields would still be useful. I'd be fine suggesting the `entropy` and `chi2` calculations as fields too, at least as a first pass, in the RFC. Statistical byte calculations seem pretty common in the binary analysis-side of security.

Reply:
agree on all points. on it